lib/sse_consumer.ex

```diff
@@ -0,0 +1,150 @@
+defmodule SSEConsumer do
+  require Logger
+  use GenServer
+
+  defmodule Request do
+    @enforce_keys ~w(method url body headers)a
+    defstruct @enforce_keys
+  end
+
+  defmodule State do
+    @moduledoc false
+    @enforce_keys ~w(recipient request conn_id remaining_stream)a
+    defstruct @enforce_keys
+
+    def new(recipient, request = %SSEConsumer.Request{}) when is_pid(recipient) do
+      %__MODULE__{
+        recipient: recipient,
+        request: request,
+        remaining_stream: "",
+        conn_id: nil
+      }
+    end
+
+
+    def set_connection(%__MODULE__{conn_id: current_conn_id} = state, new_conn_id)
+    when new_conn_id != nil and is_nil(current_conn_id) do
+      %{state | conn_id: new_conn_id}
+    end
+  end
+
+  #===========================================================================
+  # Public API
+  #===========================================================================
+
+  def stream_to(recipient, request, options), do: start_link(recipient, request, options)
+
+  def start_link(recipient, request = %Request{}, options) do
+    initial_state = State.new(recipient, request)
+    GenServer.start_link(__MODULE__, initial_state, options)
+  end
+
+  #===========================================================================
+  # Callbacks
+  #===========================================================================
+
+  @impl GenServer
+  def init(state) do
+    send(self(), :connect)
+    {:ok, state}
+  end
+
+
+  @impl GenServer
+  def handle_info(:connect, state = %State{}) do
+    %Request{method: method, url: url, body: body, headers: headers} = state.request
+    Logger.info("SSEConsumer connecting to `#{url}`")
+    case HTTPoison.request(method, url, body, headers, stream_opts()) do
+      {:ok, %{id: conn_id}} ->
+        state = State.set_connection(state, conn_id)
+        receive do
+          %HTTPoison.AsyncStatus{id: ^conn_id, code: 200} ->
+            receive do
+              %HTTPoison.AsyncHeaders{id: ^conn_id, headers: _} ->
+                Logger.info("SSEConsumer connected to `#{url}`")
+                {:noreply, state}
+              after
+                100 ->
+                  Logger.warn("SSEConsumer timed out waiting for the headers")
+                  disconnect_and_die(state, :headers_timeout)
+            end
+          %HTTPoison.AsyncStatus{id: ^conn_id, code: 400} ->
+            receive do
+              %HTTPoison.AsyncHeaders{headers: _} ->
+                Logger.warn("SSEConsumer received HTTP 400")
+                disconnect_and_die(state, :http_400)
+            end
+          after
+            2_000 ->
+              Logger.warn("SSEConsumer timed out waiting for a response")
+              disconnect_and_die(state, :timeout)
+        end
+      {:error, %HTTPoison.Error{reason: reason}} ->
```

Comment:
I think there is a problem that a supervised SSEConsumer might try to restart too quickly. I think there should be a sleep step before it dies to throttle rate of restarts when supervised

Reply:
I'm not sure it should be the responsibility of the SSEConsumer to delay its dying.

I could add a configuration option to add a user-specified delay the dying if the request finishes in a non-normal fashion (anything but AsyncEnd)..

Reply:
I think we need to recreate what happens here https://github.com/paywithcurl/public-api/blob/master/public_api/lib/public_api/event_log.ex#L136 otherwise we might prevent public-api coming up

Reply:
I think it's handled in a simple fashion in the updated version: https://github.com/paywithcurl/public-api/blob/sse-consumer/public_api/lib/public_api/event_log.ex#L45-L48

Reply:
:ok_hand: